File: src/main/kotlin/ym_cosmetic/pick_perfume_be/perfume/entity/PerfumeNote.kt
package ym_cosmetic.pick_perfume_be.perfume.entity

import jakarta.persistence.*
import ym_cosmetic.pick_perfume_be.common.BaseTimeEntity
import ym_cosmetic.pick_perfume_be.note.entity.Note
import ym_cosmetic.pick_perfume_be.perfume.vo.NoteType

@Entity
@Table(name = "perfume_note")
class PerfumeNote private constructor(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "perfume_id", nullable = false)
    val perfume: Perfume,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "note_id", nullable = false)
    val note: Note,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: NoteType
) : BaseTimeEntity() {

    companion object {
        fun create(
            perfume: Perfume,
            note: Note,
            type: NoteType
        ): PerfumeNote {
            return PerfumeNote(
                perfume = perfume,
                note = note,
                type = type
            )
        }
    }
}

File: perfume-importer/src/entities/PerfumeNote.ts
import { Entity, ManyToOne, Enum, type EnumType } from "@mikro-orm/core";
import { BaseEntity } from "./BaseEntity";
import { Perfume } from "./Perfume";
import { Note } from "./Note";

export enum NoteType {
  TOP = "top",
  MIDDLE = "middle",
  BASE = "base",
}

@Entity()
export class PerfumeNote extends BaseEntity {
  @ManyToOne(() => Perfume)
  perfume!: Perfume;

  @ManyToOne(() => Note)
  note!: Note;

  @Enum({ items: () => NoteType, type: "string" })
  type: NoteType = NoteType.TOP;

  constructor(perfume: Perfume, note: Note, type: NoteType = NoteType.TOP) {
    super();
    this.perfume = perfume;
    this.note = note;
    this.type = type;
  }
}
